### DISCUSSION TITLE
Relaxing `--exact-reachability-metadata` ?

### DISCUSSION THREAD

**Discussion by zakkak** [oracle/graal#10264] [General] [answered]
Hello,

At the moment, `--exact-reachability-metadata` requires all resources being accessed at run-time to be registered at build-time. This is a great way to detect if the code is trying to access some resource that has been mistakenly left out and is not available at run-time. However, it also results in errors when using libraries that try to load resources based on "trial and error", e.g. to find configuration files under different folders like the user's home directory, the current directory, etc.

Having to register all the potential resources seems too tedious. I wonder if an option to exclude specific code-paths, classes, or packages from the `exact-reachability-metadata` checks would make sense. E.g., for resources we could allow unregistered accesses from package `io.smallrye.config`.

cc @vjovanov

**Comment by nreid260** [comment]
I'm a bit unclear what the request is. From a technical standpoint, native-image must know about a resource at buildtime to be able to package it into the binary for runtime. Otherwise, it's up to the program to read the data from an external source (e.g. file, database).

In previous versions of native-image, it was possible to manually specify sets of resources using a regex. I *think* that went away with the `[reachability-metadata.json` format. Is this the same as what you're asking for?

https://www.graalvm.org/latest/reference-manual/native-image/metadata/ also mentions `--exact-reachability-metadata=<comma-separated-list-of-packages>` which sounds similar to the FR above.

**Reply by zakkak** [reply]
> I'm a bit unclear what the request is. From a technical standpoint, native-image must know about a resource at buildtime to be able to package it into the binary for runtime. Otherwise, it's up to the program to read the data from an external source (e.g. file, database).

Correct, the issue manifests when the program tries to access some resource at runtime that we don't necessarily need and thus didn't register. 

One could argue that the right thing to do in such cases is to still register the resource and just don't provide it at build-time (in order to not include it in the binary). However, this becomes pretty tedious when a library tries to access a resource by trying different names and/or different paths. E.g. a library might try to load files matching a pattern like `<profile>-config.properties`, where `<profile>` comes from a set of user defined profiles.

> In previous versions of native-image, it was possible to manually specify sets of resources using a regex. I think that went away with the [reachability-metadata.json format. Is this the same as what you're asking for?

Maybe. A regex/wildcard/glob would work-around the issue by essentially registering all `*-config.properties` but in more complex scenarios it could result in registering more things than intended.

FWIW `reachability-metadata.json` might not support regular expressions but it [supports globs](https://github.com/oracle/graal/blob/a701086382c5a171c5450f8f074227996f42b8ac/docs/reference-manual/native-image/assets/reachability-metadata-schema-v1.0.0.json#L83-L86) so it is still possible to achieve what you suggest.

> https://www.graalvm.org/latest/reference-manual/native-image/metadata/ also mentions `--exact-reachability-metadata=<comma-separated-list-of-packages>` which sounds similar to the FR above.

Correct, but the intention is to ultimately make `--exact-reachability-metadata` the default, meaning that we would need the exact opposite -- a way to exclude packages -- in the future.

**Comment by vjovanov** [comment]
@zakkak I am also considering to relax the current setting. I feel we have two major culprits here: 
1) **Resources.** Here one can use the `folder/**` notation to allow all resources from a single folder. I feel we don't need to relax that. 
2) **Reflection.** Here the most problematic is `Class.forName` as that is what people use to scan for libraries. I see three possible options:
     1. To scan all the classes on the classpath and if `Class.forName` is used on any of the existing classes we throw a regular exception.  Pros: easy to use. Cons: extra data in the binary, requires scanning the whole classpath, can be problematic for security and obfuscation ass all class names are included. 
     2. To allow wildcards for reflection as well `io.smallrye.config.**`. Pros: relatively easy (still misses individual queries), Cons: Includes all the types (no code) in the image.
     3. To leave the semantics of  `Class.forName` as is but: modify the error message, and allow a mode where the special exceptions are thrown (or printed). 

I am not in favor of adding a special flag to make things work (for debugging it is OK), as this moves Native Image away from the Java semantics. My plan is to experiment with these three approaches in the next 2 months and then we can decide what is easiest to support and use. @zakkak what are your thoughts?

**Reply by zakkak** [reply]
> Resources. Here one can use the folder/** notation to allow all resources from a single folder. I feel we don't need to relax that.

That indeed works well for cases where we expect the various resources to reside on a single folder (which is known at build-time), however there are cases where different folders are visited as well and to make matters worse these folders might be unknown at build-time. cc @radcortez

> Reflection. Here the most problematic is Class.forName as that is what people use to scan for libraries.

+1, I have also seen the pattern described in https://github.com/jobrunr/jobrunr/issues/1101#issuecomment-2346460357 used in more than one places as well. Users interested in a method, instead of explicitly trying to reflectively get the specific method, load the class and invoke `getDeclaredMethods` to iterate over them and see if the method they need is present. Without `--exact-reachability-metadata` they were able to get away with registering only the methods they are interested in and `getDeclaredMethods` would return only those without any exceptions/warnings. This is no longer possible though, requiring all methods to be registered...

> To scan all the classes on the classpath and if `Class.forName` is used on any of the existing classes we throw a regular exception. Pros: easy to use. Cons: extra data in the binary, requires scanning the whole classpath, can be problematic for security and obfuscation ass all class names are included.

I guess there is a typo here and the exception would be thrown when the `Class.forName` is used on any of the **non** existing classes. Right?

> To allow wildcards for reflection as well `io.smallrye.config.**`

Packages without wildcards might do the trick as well.

It's not clear to me what this would do though? Would it allow us to reflectively access any Class in `io.smallrye.config.**`? 

What we are more interested in is allowing any Class/Method under `io.smallrye.config.**` to use reflection even on unregistered elements. E.g. invoking `Class.forName("NotRegisteredClass")` from `io.smallrye.config.MyClass.loadNotRegisteredClass()` would throw a regular ClassNotFoundException

> To leave the semantics of Class.forName as is but: modify the error message, and allow a mode where the special exceptions are thrown (or printed).

Don't we already have that with `-XX:MissingRegistrationReportingMode=...`?

> I am not in favor of adding a special flag to make things work (for debugging it is OK), as this moves Native Image away from the Java semantics. 

I was thinking about extending the `reachability-metadata` or some new API that can be used in a Feature instead of a flag. E.g. for `reachability-metadata` we could have something like:

```json
{
  "reflection": [
    {
      "type": "io.smallrye.config.MyClass",
      "uncheckedReflectiveAccesses": true
    }
  ],
  "resources": [
    {
      "type": "io.smallrye.config.MyClass"
      "uncheckedReflectiveAccesses": true
    }
  ]
}
```

which would result in ignoring unregistered accesses from `io.smallrye.config.MyClass`.

> My plan is to experiment with these three approaches in the next 2 months and then we can decide what is easiest to support and use.

I am afraid 1 and 2 are going to be too expensive (in terms of bloat).

**Reply by radcortez** [reply]
How about `java.lang.ClassLoader#getResources`?

To load configuration, we use this API to search for optional resources, since it either returns an enum with the found resources or an empty one if no resources are found. Right now, if you try with `-H:+ThrowMissingRegistrationErrors` you do get the `MissingResourceRegistrationError`. It feels a bit awkward because it feels that it is changing the method behavior.

In Quarkus, we discover these resources during build time, and byte code generate a programmatic representation of the files, so we don't need the resources. We leave the method call from the library to return an empty enumeration, which works fine. We could register the resources, but the library has semantics where additional resources can be added at runtime, which are searched of both classpath, and file system. Of course, during native image, you can't add resources in the classpath, but the library is not aware of that, and it searches both paths. In this case, we cannot register the resources, because they are unknown to us during the build.

I understand the decision for APIs as `Class.forName` or equivalent that looks for a specific resource. Maybe `java.lang.ClassLoader#getResources` shouldn't do this verification?

**Reply by vjovanov** [reply]
The awkwardness comes with a benefit that you know that the native execution will behave exactly the same as the JVM. I would always trade a few more entries in the JSON file (that are explicitly mentioned in the error), for hours spent in debugging with `gdb`.

How many resource entries do you need to add in this case?  Could you please share a link to the pattern that is used here? Have you tried using the resource patterns with `**` and `*`?

**Reply by radcortez** [reply]
In our case, we support the resource as a runtime configuration:
https://github.com/smallrye/smallrye-config/blob/main/implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceLoader.java#L97-L162

For example, you can add `smallrye.config.locations=config.properties`, which will look for `config.properties` in both the classpath and working directory. This property is not known at build time, so we cannot register these resources.

Obviously, the classpath lookup will not discover any resources in native mode, but this is fine. I can work around the error by providing a substitution, but I feel that `java.lang.ClassLoader#getResources` shouldn't fail. While I agree it would behave the same if you have resources / registered resources, it will not behave the same when no resources are found in JVM mode. The JVM will just return an empty collection while you get an error in native mode.

**Comment by vjovanov** [comment]
> +1, I have also seen the pattern described in https://github.com/jobrunr/jobrunr/issues/1101#issuecomment-2346460357 used in more than one places as well.
This pattern is handled with `type` which includes all metadata for reflection. Only invocations need to be registered.

> I guess there is a typo here and the exception would be thrown when the Class.forName is used on any of the non existing classes. Right?
Sorry for the imprecision, by regular I meant `ClassNotFoundException`. We are in agreement here I think. 

Regarding bloat, I find that the biggest problem here are not the application classes, but the JDK classes. In a simple `Hello, World!` there would be 10s of thousands of types which translates to more than 1 MB. For something like PetClinic this would not be a problem as there is a relatively small number of types relative to binary size. Back of the envelope calculation: Spring PetClinic has ~40000 unreached types on the classpath, each type can be stored in a trie structure, so we need roughly only the unqualified class name extra in the heap. `40000 * 10 (class name) * 2 (bytes per char)` is roughly 1 MB which is little compared to PetClinic that has ~200 MB.

> It's not clear to me what this would do though? Would it allow us to reflectively access any Class in io.smallrye.config.**?

Yes it would include all of these types to be available for `Class.forName`.

> Don't we already have that with `-XX:MissingRegistrationReportingMode=...`?

We do, exactly, just the default exception for `Class.forName` would be a `ClassNotFoundException` with a proper error message about missing registration.

> I was thinking about extending the reachability-metadata or some new API that can be used in a Feature instead of a flag. E.g. for reachability-metadata we could have something like:

When you boil it down, this is a workaround for a lack of `MissingReflectionRegistrationError` in the JDK. If this existed, and even better if it was a supertype of (or has a common super type with) the `ClassNotFoundException`, the libraries would be able to catch that exception and simply ignore it. 

Maybe we could throw `ReflectiveOperationException` instead of `ClassNotFoundException`? 

This way libraries would be able to handle this and we don't need extra JSON. In the future if a `MissingReflectionRegistrationException`  is added in between  `ReflectiveOperationException` and `ClassNotFoundException` (or as a sibling to `ClassNotFoundException`) all the code would continue to work as expected.

> Note: throughout this discussion I am trying to minimize the amount of JSON. My goal is that all of the semantics can be eventually expressed in user code without any external files.

**Reply by zakkak** [reply]
> This way libraries would be able to handle this and we don't need extra JSON. In the future if a MissingReflectionRegistrationException is added in between ReflectiveOperationException and ClassNotFoundException (or as a sibling to ClassNotFoundException) all the code would continue to work as expected.

In theory I agree, but this approach assumes that libraries will be willing to include code specifically for handling GraalVM native-image compilation, which is not always the case. Metadata (like substitutions) allow us to work around this and support libraries that are not willing to include GraalVM specific code.

**Comment by vjovanov** [comment]
> which would result in ignoring unregistered accesses from `io.smallrye.config.MyClass`

We could make a flag for this, but I would leave it as experimental for the time the third-party libraries don't adopt the new semantics.

**Comment by nreid260** [comment]
Have you considered branching the logic of the original code on the `org.graalvm.nativeimage.imagecode` system property? That way you could ignore codepaths that never work after graalification, rather than trying to stub them to work when resources are gone.

We created an enum around the possible values for convenience.

**Reply by zakkak** [reply]
@nreid260 if I get this right, you suggest adding a check to guard resource accesses, e.g.

```
if (!inNative) {
  accessResource
}
```

That's indeed an option when you have write access (or can convince the maintainers to merge this kind of change), i.e. this approach assumes that libraries will be willing to include code specifically for handling GraalVM native-image compilation, which is not always the case. As a result, in some cases we would need to introduce substitutions to apply this.

**Reply by nreid260** [reply]
Yeah this is the basic idea. It's fairly natural to me that _some_ codepaths which work for JVM and need to be changed for native.

For libraries you don't control, you could do a little bytecode rewriting (e.g. with ASM) to insert the needed logic. Of course this takes effort, but possibly less effort than other options.

I also wonder if library authors would be sympathetic to simpler resource handling logic even outside the context of graalification. They might support args like "skip loading external config files" or whatever.

**Reply by vjovanov** [reply]
Yes, I full agree with the argument here. I would always strive to have a simple solution that is the same both with native and on the JVM. For example, in [radcortez](https://github.com/radcortez)'s example above it would be nice to look only on the file path and skip resources lookup altogether. 

But if this is not possible for some reason, we can always make a condition on `org.graalvm.nativeimage.imagecode`. As long as we have only a small bunch of these places in the ecosystem, it should be OK.

**Reply by zakkak** [reply]
We discussed this with @radcortez and it's still problematic. Imagine the following scenario:

The library allows users to load arbitrary files either from the classpath or the filesystem. The user defines the file name through a property called `my.file.property`. This property can be set at run-time.

In JVM mode everything works as expected. The user sets `my.file.property` to `my.config` and the library looks up the for files named `my.config` in the classpath or the filesystem. If `my.config` is not present in the classpath `getResources` returns nothing.

In native mode the value of `my.file.property` is not known at compile-time so there is no way to register it's value for run-time access, as a result the application fails when looking up `my.config` in the classpath.

Substituting the logic and avoiding the classpath lookup works around the issue, but behaves differently than the JVM mode, since in native mode the classpath is never searched. That sounds fine since even in the current state that allows searching in the classpath the behavior would be different if the resource was not registered.

The problem arises if an application developer using the library registers some resources, e.g. `a.config` and `b.config`. At the moment, this allows the application users to set `my.file.property` to any of these two values and everything should work just like in JVM-mode. However, if we substitute the code to avoid classpath lookups in native mode this will no longer work, the users won't get an exception but they will also not be able to use the registered `a.config` and `b.config` resources. Without the substitutions they will be able to use the registered `a.config` and `b.config` resources, but passing any other resource name to `my.file.property` will result in a `MissingRegistrationException`.

So far the only backwards compatible approach we could come up is to substitute the library code to introduce a try-catch that will handle the `MissingRegistrationException` (assuming the upstream library is not willing to include graalvm-specific code). This however comes with the maintenance cost of keeping the substitution up to date with the upstream code.

**Comment by vjovanov** [comment]
In general, to make an informed decision it would be good to give examples of how many entries relative to all metadata entries need to be added for the project to work. 

If the number of entries is relatively small compared to the size of the project (e.g. <1-2% of all entries), I feel the benefits of strict semantics outweigh the cost of maintaining the extra entries (especially because the missing entries are pretty easy to find).

**Reply by zakkak** [reply]
It looks like there are cases that the main issue is not the number of entries but the difficulty of figuring out what that entries should be. One example is given in https://github.com/oracle/graal/discussions/10264#discussioncomment-11604091 and I just stumbled upon another interesting case, this time in the JDK itself. 

The issue in this case is not related to resources, but constructors. In `sun.net.www.protocol.http.HttpURLConnection#getInputStream0` and `sun.net.www.protocol.http.HttpURLConnection#getOutputStream0` the JDK is [using reflection to instantiate a "cached" exception](https://github.com/openjdk/jdk/blob/b4cd3b12967688d78028f548db76d24cfe654614/src/java.base/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L1740-L1756) (stored in `rememberedException` and `rememberedOutException` respectively) which results in a `MissingReflectionRegistrationError`. Without `--exact-reachability-metadata` the application runs as expected since the constructor is available in the image (otherwise the exception wouldn't be instantiated in the first place) even though it was not explicitly registered for reflection. In this as well as in similar cases it's not trivial to detect all the possible exceptions that might get thrown in order to register them. One needs to replicate the work done by the static analysis to identify all the reachable code paths that may throw an exception that will be then stored in `rememberedException` or `rememberedOutException`.

Are there any suggestions on how to approach this issue?

**Comment by vjovanov** [comment, marked as answer]
After a few iterations we have concluded the following: 

1. `--exact-reachability-metadata` will remain very strict in order to guarantee that all entries are covered by the metadata. This includes resource lookups and `Class.forName` on non-existing classes. This flag should be used by the frameworks and libraries to make sure that they have covered all possible entries in the code and that their users don't have to do it. 

2. Most of the functionality `--exact-reachability-metadata` currently covers will be promoted to the default behavior (e.g., throwing a `MissingReflectionRegistrationError` when calling `Class.getDeclaredMethods()` on an unregistered class). With this, the end user will not care about negative queries (unless when debugging), but they will get informative errors when their metadata is incomplete.

3.  `--exact-reachability-metadata` will likely be promoted to a run-time flag so the image doesn't have to be rebuilt to use it. I expect this to happen in GraalVM for JDK 26.

In the example you provided with resources, if there is a test that passes a system property that leads to reading the resource in many different places, I feel it is OK that this test also provides the metadata for those places. The end users will eventually not need to do it, but the frameworks will have to go this extra step. If really cumbersome, one can always use wildcards in `glob` to cover all of such entries.

**Reply by zakkak** [reply]
@vjovanov can you please clarify what functionality of `--exact-reachability-metadata` is expected to not get promoted to the default behavior. For instance, from the above I understand that resource access checks won't be promoted (at least according to the current plan).

**Reply by vjovanov** [reply]
Based on our experience in the community, the things that will not be promoted to default are special exceptions for `Class.forName` and resource accesses (not resource bundles). These functions are often used to "poke" the class path and can require too much metadata or input-dependent metadata.

End users will always be able to enable the strict mode to keep their metadata strict or to debug when the applications don't work. Framework and library authors should always use the strict mode to make sure all metadata is accounted for.
